jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,718 @@
+ecma_call_builtin_executor (ecma_object_t *executor_p, /**< the executor object */
+                            ecma_value_t resolve_func, /**< the resolve function */
+                            ecma_value_t reject_func) /**< the reject function */
+{
+  ecma_string_t *str_0 = ecma_new_ecma_string_from_uint32 (0);
+  ecma_string_t *str_1 = ecma_new_ecma_string_from_uint32 (1);
+  ecma_string_t *str_2 = ecma_new_ecma_string_from_uint32 (2);
+  /* 2. String '0' indicates [[Capability]] of the executor. */
+  ecma_value_t capability = ecma_op_object_get (executor_p, str_0);
+  /* 3. String '1' indicates [[Resolve]] of the capability. */
+  ecma_value_t resolve = ecma_op_object_get (ecma_get_object_from_value (capability), str_1);
+
+  if (resolve != ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED))
+  {
+    ecma_free_value (resolve);
+    ecma_free_value (capability);
+    ecma_deref_ecma_string (str_0);
+    ecma_deref_ecma_string (str_1);
+    ecma_deref_ecma_string (str_2);
+
+    return ecma_raise_type_error (ECMA_ERR_MSG ("'resolve' function should be undefined."));
+  }
+
+  /* 4. String '2' indicates [[Reject]] of the capability. */
+  ecma_value_t reject = ecma_op_object_get (ecma_get_object_from_value (capability), str_2);
+
+  if (reject != ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED))
+  {
+    ecma_free_value (reject);
+    ecma_free_value (capability);
+    ecma_deref_ecma_string (str_0);
+    ecma_deref_ecma_string (str_1);
+    ecma_deref_ecma_string (str_2);
+
+    return ecma_raise_type_error (ECMA_ERR_MSG ("'reject' function should be undefined."));
+  }
+
+  /* 5. */
+  ecma_op_object_put (ecma_get_object_from_value (capability),
+                      str_1,
+                      resolve_func,
+                      false);
+  /* 6. */
+  ecma_op_object_put (ecma_get_object_from_value (capability),
+                      str_2,
+                      reject_func,
+                      false);
+
+  ecma_free_value (capability);
+  ecma_deref_ecma_string (str_0);
+  ecma_deref_ecma_string (str_1);
+  ecma_deref_ecma_string (str_2);
+
+  return ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+} /* ecma_call_builtin_executor */
+
+/**
+ * Create a PromiseCreateResovingFucntions.
+ *
+ * See also: ES2015 25.4.1.3
+ *
+ * @return pointer to the resolving functions
+ */
+ecma_promise_resolving_functions_t *
+ecma_promise_create_resolving_functions (ecma_object_t *object_p) /**< the promise object */
+{
+  /* 1. */
+  ecma_value_t already_resolved = ecma_op_create_boolean_object (ecma_make_boolean_value (false));
+
+  ecma_string_t str_promise;
+  ecma_string_t str_already_resolved;
+  ecma_init_ecma_magic_string (&str_promise, LIT_INTERNAL_MAGIC_STRING_PROMISE);
+  ecma_init_ecma_magic_string (&str_already_resolved, LIT_INTERNAL_MAGIC_STRING_ALREADY_RESOLVED);
+  /* 2. */
+  ecma_object_t *resolve_p;
+  resolve_p = ecma_op_create_external_function_object ((ecma_external_pointer_t) ecma_promise_resolve_handler);
```

Comment:
Are these functions can be called by users? They seem internal based on their name. I would prefer just directly call them.